### ISSUE TITLE
Output channels select option on MacOS

### ISSUE THREAD

**Issue by biomassa** [EmissionControl2/EmissionControl2#69]
It would be great if at some point down the path some sort of simple audio output selection is added. 
At this time when using a multichannel audio interface (MOTU 896HD in this case) Emissions Control seems to default to main outputs without a way to change that. It would be nice to be able to select the output pair to allow for more flexible routing (e.g. send EC's output to modular for futher processing).
Thanks!

**Comment by jackkilgore** [comment]
Hello, if you go to "Audio -> AudioOutput -> Stop" a dropdown menu will then appear where you can select the desire output device. Is this what you need?

**Comment by biomassa** [comment]
Thanks for your reply! However, I'm talking about the ability to select a different pair of outputs on multichannel output devices — like the MOTU896HD that has 8 outputs. 
in "Audio -> AudioOutput -> Stop" I can select only the device, but not its output pair.

Thanks!

**Comment by jackkilgore** [comment]
Ohhhhhhh, I SEE, that makes a lot of sense. I will look into that, thanks for the suggestion!

**Comment by rodneydup** [comment]
Ah yes this is a classic MOTU situation I've run into a lot. But it would be useful for other multichannel interfaces as well. In general it would be good to support selecting output channels. This actually might be fairly simple using some features built into Allolib. @mantaraya36, am I right?

**Comment by mantaraya36** [comment]
Yes, it's simple in the code, just use the index to the channel when you
get the buffer to write to. It will take more time to make a gui for this.
Perhaps a quick solution is to add an offset value, so if you want to use
channels 6 and 7, the index offset is 5 (the GUI should add one to this so
it shows 6).

If you do this with a Parameter, you probably want to get the value outside
the sample loop for efficiency.

Cheers,
Andres

**Comment by jackkilgore** [comment]
This is a question for @mantaraya36 . I have been working on this issue, but I have come up against a problem with the PolySynth class. When I call .render on it, it writes all the grains to some internal buffer that then writes those buffers to the audio output. Under the hood, the PolySynth class seems to make, lets say two channels with index [0,1] and then implicitly writes **those to the corresponding indices** on the audio out, so [0,1]. However, what I really want to happen is that it has internal buffers with indices [0,1] that can be arbitrarily written to any available output channel. 
So if I had an audio device with 16 outs, I might want to map 
     PolySynth::indicies : [0,1] -> AudioOut::indices : [3,4]

**Comment by mantaraya36** [comment]
I have just added a setChannelMap() function to PolySynth. Completely
untested and whipped up in a rush, but should do what you need while
avoiding having to copy the buffers manually.

**Comment by jackkilgore** [comment]
Thanks, Andres, I shall try it out this weekend!

**Comment by jackkilgore** [comment]
Works like a charm, this feature + the other issues with audio devices should be fixed on jkUserSpecifiedOut. If anyone wants to pull this branch and see if it works for them before I merge into main, that would be helpful!

**Comment by rodneydup** [comment]
Seems to run on Linux fine but there are a couple curious things it caused with JACK connection kit.  First one isn't a big deal but on the JACK connection graph, EC2 connects to all available ports on the selected device, but internally only sends the audio to the two outputs specified. Just looks a bit messy. I would expect it to the necessary ports ideally.

Second, EC2 now has inputs on the JACK graph. I had gotten rid of them before, but something you changed seems to have brought them back. The EC2 inputs of course don't do anything but again it just looks messy.

I can take a look at these things in the next week or two probably but right now I have some tight deadlines. In the meantime, maybe someone else already has an idea on how to fix this.

**Comment by jackkilgore** [comment]
@rodneydup If you can look at it that would be very helpful. We should meet sometime when the quarter ends to kind of reset and figure out what we want to tackle.

**Comment by jackkilgore** [comment]
@mantaraya36 The reason that the above is happening to @rodneydup is this:

Let's say I have a device X with 16 output channels and I want EC2 to use channels 15 and 16.
Ideally, I can just open up two channels corresponding to X's 15 and 16th channel.
However, the function `al::AudioIO::channels(int num, bool forOutput); ` forces us to open all channels from 1-16 before we can use the 15th and 16th channel. This is similar to the problem I had with the PolySynth class. Can we perhaps get a similar fix for the AudioIO class?

**Comment by rodneydup** [comment]
![image](https://user-images.githubusercontent.com/35479107/107161204-26e5a980-6969-11eb-963e-9d1e6f449c3c.jpeg)

This is what it looks like in JACK, for example, when I use a multi-channel interface. Desired behaviour is to only show two outputs from EC2 and no inputs at all.

**Comment by mantaraya36** [comment]
You could add on option in the GUI that sets audioIO().channels(int num, bool forOutput);

This determines the number of channels that are open by the audio device. If you want to have an offset, then you still have to open all the previous unused channels, and then offset the writing of the buffers. I don't think there is an easy solution to offset the channels opened through RtAudio. Perhaps I'm worng...

**Comment by ricknance** [comment]
A related issue on OS X. When recording it always uses the max number of channels available on the interface. The RME has 16, so regardless of just using 2 channel files, it writes to 14 channel. Kind of minor, but if you're going into it anyway, it might be useful to be able to choose that as well
